### PR TITLE
Add getversioninfo

### LINKED ISSUE #930: Add RPC `getversioninfo`
Currently, any version info from the node is provided by `getnetworkinfo`. This includes a numerical string, a string and the protocol version carried over from the bitcoin core. They are also ambiguous as they don't necessarily have to reflect the network but rather the capability of the node. 

With DfTx future versioning, introduction of a dedicated `getversioninfo` can provide a singular, friendly RPC call that can provide all info relevant to versions of both the client and network, including individual DfTx versions supported by the client.  

Expected spec:

```
{
  node: {
    "name": "DeFiChain",
    "version": "2.2.1",
    "numericVersion": 2020100,
    "fullVersion": "DeFiChain:2.2.1",
    "userAgent": ".. from our UA string (the one our client exposes ourself as..)",
  },
  rpc: { 
   version: ,
   min: ,
   max: ,
  },
  protocol: { 
   version: , 
   min: ,
   tx: { 
     ... 
   }
  },
  storage: {
    ... furthest migration, we support from ..
  },
  spv: { 
    btc: { 
      protocolVersion: ,
      minProtocolVersion: ,
      userAgent:,
    },
  }
}
```

### PR DESCRIPTION
**Pull request by mambisi**
#### What kind of PR is this?:
/kind feature

#### What this PR does / why we need it:
Add `getversioninfo` RPC 
```
defi-cli getversioninfo
{
  "node": {
    "name": "DeFiChain",
    "version": "2.2.1",
    "numericVersion": 2020100,
    "fullVersion": "DeFiChain:2.2.1",
    "userAgent": "/DeFiChain:2.2.1/"
  },
  "protocol": {
    "version": 70022
  }
}


```

#### Which issue(s) does this PR fixes?:
Fixes #930

### CODE REVIEW

File: src/rpc/net.cpp [hunk 1]
```diff
@@ -511,7 +511,30 @@ static UniValue getnetworkinfo(const JSONRPCRequest& request)
...
+                       "  \"protocolversion\": xxxxx,              (numeric) the protocol version\n"
+                       "}\n"                                                                                                                                                                                                                                                      "  \"warnings\": \"...\"                    (string) any network and blockchain warnings\n"
+                                                                                                                                                                                                                                                                                      "}\n"
+               },
+               RPCExamples{
+                       HelpExampleCli("getversioninfo", "")
+                       + HelpExampleRpc("getversioninfo", "")
+               },
+    }.Check(request);
+    LOCK(cs_main);
```
Comment (prasannavl) [inline src/rpc/net.cpp:531]: Don't need to lock the chain state here.

**Comment by prasannavl** [comment]
What do you think about formatting it like this: 

```
{ 
 client: { 
    name: `DeFiChain`,
    version: `2.2.1`,
    numericVersion: 2020100,
    fullVersion: `DeFiChain:2.2.1`,
    protocol: 70022,
 },
 ... 
 // use for txs later.
 .. 
}
```

File: src/rpc/net.cpp (hunk 1 above)
Comment (mambisi) [reply src/rpc/net.cpp:531]: Okay

**Comment by bvbfan** [comment]
Probably `Satoshi` could be `DefiChain`

**Review by hidiego — APPROVED** [review]

**Review by Jouzo — APPROVED** [review]

**Review by Jouzo — DISMISSED** [review]

**Comment by prasannavl** [comment]
Adding for future (for 2.2.2, let's just go with node)

```
{
  node: {
    "name": "DeFiChain",
    "version": "2.2.1",
    "numericVersion": 2020100,
    "fullVersion": "DeFiChain:2.2.1",
    "userAgent": ".. from our UA string (the one our client exposes ourself as..)",
  },
  rpc: { 
   version: ,
   min: ,
   max: ,
  },
  protocol: { 
   version: , 
   min: ,
   tx: { 
     ... 
   }
  },
  storage: {
    ... furthest migration, we support from ..
  },
  btc: { 
    protocolVersion: ,
    minProtocolVersion: ,
    userAgent:,
  }
}
```

**Review by hidiego — APPROVED** [review]

**Review by Jouzo — DISMISSED** [review]

File: src/rpc/net.cpp [hunk 2]
```diff
@@ -511,7 +511,44 @@ static UniValue getnetworkinfo(const JSONRPCRequest& request)
...
+                       "      \"fullVersion\": \"/DefiChain:x.x.x/\",     (string) the server subversion string\n"
+                       "      \"protocol\": xxxxx,                      (numeric) the protocol version\n"
+                       "   }\n"
+                       "  ...\n"
+                       "}"
+               },
+               RPCExamples{
+                       HelpExampleCli("getversioninfo", "")
+                       + HelpExampleRpc("getversioninfo", "")
+               },
```
Comment (prasannavl) [inline src/rpc/net.cpp:533]:
```suggestion
    RPCHelpMan{"getversioninfo",
               "Returns an object containing various version info related to the node and protocol.\n",
               {},
               RPCResult{
                       "{\n"
                       "  \"node\": {\n"
                       "      \"name\": DeFiChain                       (string) Node name\n"
                       "      \"version\": \"xxxxx\",                   (string) Node version string\n"
                       "      \"numericVersion\": xxxxx,                (number) Node numeric version\n"
                       "      \"fullVersion\": \"DefiChain:x.x.x\",     (string) Full node version string including name and version\n"
                       "      \"userAgent\": \"/DefiChain:x.x.x/\",     (string) P2P user agent string (subversion string conforming to BIP-14)\n"
                       "   },\n"
                       "  \"protocol\": {\n"
                       "      \"version\": \"xxxxx\",                   (number) Operating protocol version\n"
                       "      \"min\": \"xxxxx\",                       (number) Minimum protocol that's supported by the node\n"
                       "   },\n"
                       "  ...\n"
                       "}"
               },
               RPCExamples{
                       HelpExampleCli("getversioninfo", "")
                       + HelpExampleRpc("getversioninfo", "")
               },
```

File: src/net.h [hunk 3]
```diff
@@ -532,6 +532,10 @@ extern bool g_relay_txes;
 /** Subversion as sent to the P2P network in `version` messages */
 extern std::string strSubVersion;
 
+extern std::string strVersion;
+
+extern std::string strFullVersion;
+
```
Comment (prasannavl) [inline src/net.h:538]: Remove.

```suggestion
```

File: src/net.cpp [hunk 4]
```diff
@@ -88,6 +88,8 @@ CCriticalSection cs_mapLocalHost;
 std::map<CNetAddr, LocalServiceInfo> mapLocalHost GUARDED_BY(cs_mapLocalHost);
 static bool vfLimited[NET_MAX] GUARDED_BY(cs_mapLocalHost) = {};
 std::string strSubVersion;
+std::string strVersion;
+std::string strFullVersion;
```
Comment (prasannavl) [inline src/net.cpp:92]: Remove these. No point keeping strings around globally that aren't used else where and extremely cheap to derive.

```suggestion
```

File: src/init.cpp [hunk 5]
```diff
@@ -1401,6 +1401,8 @@ bool AppInitMain(InitInterfaces& interfaces)
         uacomments.push_back(cmt);
     }
     strSubVersion = FormatSubVersion(CLIENT_NAME, CLIENT_VERSION, uacomments);
+    strVersion = FormatVersion(CLIENT_VERSION);
+    strFullVersion = FormatFullVersion(CLIENT_NAME, CLIENT_VERSION);
```
Comment (prasannavl) [inline src/init.cpp:1405]: Remove these. No point keeping strings around globally that aren't used else where and extremely cheap to derive.

```suggestion
```

File: src/clientversion.cpp [hunk 6]
```diff
@@ -105,3 +105,10 @@ std::string FormatSubVersion(const std::string& name, int nClientVersion, const
     ss << "/";
     return ss.str();
 }
+
+std::string FormatFullVersion(const std::string& name, int nClientVersion)
+{
+    std::ostringstream ss;
+    ss << name << ":" << FormatVersion(nClientVersion);
+    return ss.str();
+}
```
Comment (prasannavl) [inline src/clientversion.cpp:114]: Remove, no need. It can be simple + instead the rpc call when needed.

File: src/clientversion.h [hunk 7]
```diff
@@ -44,8 +44,10 @@ static const int CLIENT_VERSION =
 extern const std::string CLIENT_NAME;
 extern const std::string CLIENT_BUILD;
 
+std::string FormatVersion(int nVersion);
 std::string FormatFullVersion();
 std::string FormatSubVersion(const std::string& name, int nClientVersion, const std::vector<std::string>& comments);
+std::string FormatFullVersion(const std::string& name, int nClientVersion);
```
Comment (prasannavl) [inline src/clientversion.h:50]: Remove.

File: src/clientversion.h [hunk 8]
```diff
@@ -44,8 +44,10 @@ static const int CLIENT_VERSION =
 extern const std::string CLIENT_NAME;
 extern const std::string CLIENT_BUILD;
 
+std::string FormatVersion(int nVersion);
 std::string FormatFullVersion();
```
Comment (prasannavl) [inline src/clientversion.h:48]: We can use this opportunity to change this into something more meaningful. The current terminology introduces a lot of confusion - Version vs Subversion vs FullVersion. (Full version lacks clarity on what exactly is full and opens up for mistakes in the future). 

`FormatFullVersion` appears to be version with suffix (like branch names). How about renaming this to  `FormatVersionAndSuffix` instead?

**Review by prasannavl — CHANGES_REQUESTED** [review]

**Comment by prasannavl** [comment]
In addition to the above, also make sure UA String is exposed correctly, corresponding to the BIP spec.

**Comment by prasannavl** [comment]
I'm moving this to next version. Let's also add `rpc` and `btc` into it.

File: src/rpc/net.cpp (hunk 2 above)
Comment (mambisi) [reply src/rpc/net.cpp:533]: Okay

File: src/net.h (hunk 3 above)
Comment (mambisi) [reply src/net.h:538]: Okay

File: src/net.cpp (hunk 4 above)
Comment (mambisi) [reply src/net.cpp:92]: :+1:

File: src/clientversion.h (hunk 8 above)
Comment (mambisi) [reply src/clientversion.h:48]: Okay

File: src/clientversion.h [hunk 9]
```diff
@@ -44,7 +44,8 @@ static const int CLIENT_VERSION =
 extern const std::string CLIENT_NAME;
 extern const std::string CLIENT_BUILD;
 
-std::string FormatFullVersion();
+std::string FormatVersion(int nVersion);
+std::string FormatVersionAndSuffix();
 std::string FormatSubVersion(const std::string& name, int nClientVersion, const std::vector<std::string>& comments);
```
Comment (prasannavl) [inline src/clientversion.h:49]: While at it, let's make this clear too - change htis to `FormatUserAgentString` (Also cover rename boost tests accordingly)

File: src/rpc/rpcversion.h [hunk 10]
```diff
@@ -0,0 +1,22 @@
+#ifndef DEFI_RPCVERSION_H
```
Comment (prasannavl) [inline src/rpc/rpcversion.h:1]: I don't think we need semver for RPC. It's just going to be numeric 1, 2, 3..

I think we can remove this entirely and just hard code both min, and current to 1 for now.

Comment (mambisi) [reply src/rpc/rpcversion.h:1]: Okay

File: src/clientversion.h (hunk 9 above)
Comment (mambisi) [reply src/clientversion.h:49]: Okay

**Pull request merged**